src/main/java/fr/sncf/osrd/train/Train.java

```diff
@@ -101,6 +103,35 @@ public void scheduleStateChange(Simulation sim) throws SimulationError {
         lastScheduledEvent = clonedState.simulatePhase(this, sim);
     }
 
+
+    /** Frees all the tvd sections currently reserved by the train */
+    private void freeAllReservedTVDSections(Simulation sim) throws SimulationError {
+        var path = lastState.path;
+        var trainTailLocation = lastState.location.trackSectionRanges.getLast().getBeginLocation();
+
+        // all the TVD sections in routes we have already reached
+        var tvdSectionsOnOnceReservedRoutes = new ArrayList<TVDSectionPath>();
+        for (int i = 0; i <= lastState.requestedRouteIndex; i++)
+            tvdSectionsOnOnceReservedRoutes.addAll(path.routePath.get(i).tvdSectionsPaths);
+
+        var beforeTrainTail = true;
+        for (var currentTvdSectionPath : tvdSectionsOnOnceReservedRoutes) {
+
+            if (currentTvdSectionPath.contains(trainTailLocation))
+                beforeTrainTail = false;
+
+            // ignores the tvd sections before the start of the train (already freed)
+            if (beforeTrainTail)
+                continue;
+
+            var index = currentTvdSectionPath.tvdSection.index;
+            var tvdSection = sim.infraState.getTvdSectionState(index);
+
+            tvdSection.free(sim);
+        }
+
+    }
```

Comment:
this seems super fishy to me, we have a queue of actions points waiting for a tail interaction, why can't we just empty this queue and signal all tvdsections the train is gone?

see https://github.com/DGEXSolutions/osrd-core/blob/853d02ff907aabe3ddd003e7a3f6889f0f47e1a5/src/main/java/fr/sncf/osrd/train/TrainState.java#L43